File: python/SignValidator.py
def is_valid(signos: str) -> bool:
    stack = []

    for c in signos:
        if c in "([{":
            stack.append(c)

        elif c in ")]}":
            if not stack:
                return False

            ultimo = stack.pop()

            if not (
                (ultimo == '(' and c == ')') or
                (ultimo == '{' and c == '}') or
                (ultimo == '[' and c == ']')
            ):
                return False

    return len(stack) == 0

File: src/utils/SignValidator.java
package utils;

import java.util.Stack;

public class SignValidator {
    public boolean isValid(String signos) {
        Stack<Character> pila = new Stack<Character>();
        for (int i = 0; i < signos.length(); i++) {
            char c = signos.charAt(i);

            if (open(c)) {
                pila.push(c);
            } else if (close(c)) {
                if (pila.isEmpty())
                    return false;

                char ultimo = pila.pop();
                if (!correctSyntax(ultimo, c)) {
                    return false;
                }
            }

        }
        return pila.isEmpty();
    }

    private static boolean open(char c) {
        return c == '(' || c == '[' || c == '{';
    }

    private static boolean close(char c) {
        return c == ')' || c == ']' || c == '}';
    }

    private static boolean correctSyntax(char open, char close) {
        return (open == '(' && close == ')') ||
                (open == '{' && close == '}') ||
                (open == '[' && close == ']');
    }
}
